### PR TITLE
Leverage coalescing operators. NFC

### PR DESCRIPTION
**Pull request by RReverser**
Note that this bumps minimum engine versions that can avoid transpilation.

I believe this is not blocked on https://github.com/emscripten-core/emscripten/issues/11984 as it doesn't require Closure to emit runtime helpers.

This is mostly automated with ast-grep, bunch of custom rules, and a few manual fixups. Best viewed with "ignore whitespace changes" (https://github.com/emscripten-core/emscripten/pull/20531/files) due to indentation changes and due to first commit being #20530.

Care was taken to ensure that false-y conditions still use `||` where it might matter and only null-ish conditions use `??`.

### CODE REVIEW

**Comment by RReverser** [comment]
Sigh, does that failure mean it would require Node.js update too? I'm guessing that's less likely?

**Comment by sbc100** [comment]
This is awesome!

Can you update `test_es5_transpile` to make sure this transpiles cleanly with closure copiler?

Normally/historically when I make this kind of change I like to land the change the bumps to browser versions and updates `test_es5_transpile` (and possibly introduces just one usage of the new construct into the stdlib) first.. and then to the large scale change as a followup.   I think this helps with review and makes bisecting easier.

**Comment by sbc100** [comment]
> Sigh, does that failure mean it would require Node.js update too? I'm guessing that's less likely?

Old versions of node (just like old browser versions) should be supported via transpilation, no?

**Comment by RReverser** [comment]
> Old versions of node (just like old browser versions) should be supported via transpilation, no?

Probably. I'm not entirely sure what this `test_node_compat` job does, never looked at it before, I just see that it fails because it seems to use rather old Node.

**Comment by RReverser** [comment]
> Normally/historically when I make this kind of change I like to land the change the bumps to browser versions and updates `test_es5_transpile` (and possibly introduces just one usage of the new construct into the stdlib) first.. and then to the large scale change as a followup. I think this helps with review and makes bisecting easier.

That makes sense, probably helps with ignoring revisions too. I'm happy to split out if the PR is looking acceptable otherwise, but first would probably need to figure out this failure.

**Comment by RReverser** [comment]
I'm not even sure why I'm doing this. I don't need it for my work, but this kind of large-scale refactorings is somewhat addicting 😅

**Comment by RReverser** [comment]
...huh, those are some weird "bulk memory support is not enabled" failures. Especially given that it's couple of standalone tests that are failing - those that don't depend on JS.

**Comment by RReverser** [comment]
> Probably. I'm not entirely sure what this `test_node_compat` job does, never looked at it before, I just see that it fails because it seems to use rather old Node.

> /home/circleci/node-v10.19.0-linux-x64/bin/node /home/circleci/project/src/compiler.js /tmp/emtest_d5kxfnv8/tmpg3hc1hda.json --symbols-only

Just noticed the version, that's very old Node indeed. Why does that job use it instead of the one from emsdk which I believe is newer?

File: src/library_browser.js [hunk 1]
```diff
@@ -1022,7 +1022,7 @@ var LibraryBrowser = {
       // to queue the newest produced audio samples.
       // TODO: Consider adding pre- and post- rAF callbacks so that GL.newRenderingFrameStarted() and SDL.audio.queueNewAudioData()
       //       do not need to be hardcoded into this function, but can be more generic.
-      if (typeof SDL == 'object' && SDL.audio && SDL.audio.queueNewAudioData) SDL.audio.queueNewAudioData();
+      if (typeof SDL == 'object') SDL.audio?.queueNewAudioData?.();
```
Comment (kripken) [inline src/library_browser.js:1025]: Does `?.` work on checking if a symbol exists in the global scope? If so we could do
```suggestion
      SDL?.audio?.queueNewAudioData?.();
```
Perhaps that would run into a closure error though...

File: src/gl-matrix.js [hunk 2]
```diff
@@ -180,7 +180,7 @@ vec3.multiply = function (vec, vec2, dest) {
  * _returns {vec3} dest if specified, vec otherwise
  */
 vec3.negate = function (vec, dest) {
-    if (!dest) { dest = vec; }
+    dest ||= vec;
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/gl-matrix.js:183]: Oh I didn't realise that. I don't mind reverting. OTOH it looks like it was only imported once and then had various Closure fixups.

File: src/library_wasmfs.js [hunk 3]
```diff
@@ -542,7 +542,7 @@ FS.init();
 
   $FS_mkdir__deps: ['_wasmfs_mkdir'],
   $FS_mkdir: (path, mode) => FS.handleError(withStackSave(() => {
-    mode = mode !== undefined ? mode : 511 /* 0777 */;
+    mode = mode ?? 511 /* 0777 */;
```
[the remark this answers is not in the document]
Comment (RReverser) [reply src/library_wasmfs.js:545]: Yeah exactly, I worked under assumption that mistaking false-y values (`0`, `false`, `""`) in such contexts could be bad but `null` and `undefined` are likely intechangeable. `??` and `?.` were basically added to JS under the same assumption.

File: src/gl-matrix.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbc100) [reply src/gl-matrix.js:183]: I would probably just revert this file.. or at least document this change in the `Modifed for emscripten:` comment at the top.

**Comment by RReverser** [comment]
> Can you update `test_es5_transpile` to make sure this transpiles cleanly with closure copiler?
> 
> Normally/historically when I make this kind of change I like to land the change the bumps to browser versions and updates `test_es5_transpile` (and possibly introduces just one usage of the new construct into the stdlib) first.. and then to the large scale change as a followup. I think this helps with review and makes bisecting easier.

Extracted in https://github.com/emscripten-core/emscripten/pull/20549.

File: src/gl-matrix.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/gl-matrix.js]: Reverted.

File: src/library_wasmfs.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RReverser) [reply src/library_wasmfs.js]: Updated to use default param value in those particular functions as they're even more appropriate here and keep the undefined check.

File: test/other/test_unoptimized_code_size_strict.js.size [hunk 4]
```diff
@@ -1 +1 @@
-57018
+58622
```
Comment (sbc100) [inline test/other/test_unoptimized_code_size_strict.js.size:1]: These are unoptimized sizes so I guess its OK that they get larger... but I'm kind of surprised we don't see any code size wins for the optimized builds..  any idea why we don't?

File: src/library_html5_webgl.js [hunk 5]
```diff
@@ -31,10 +31,10 @@ var LibraryHtml5WebGL = {
...
 
     HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.alpha }}}>>2)] =
-    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.depth }}}>>2)] = 
-    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.antialias }}}>>2)] = 
-    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.premultipliedAlpha }}}>>2)] = 
-    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.majorVersion }}}>>2)] = 
+    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.depth }}}>>2)] =
+    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.antialias }}}>>2)] =
+    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.premultipliedAlpha }}}>>2)] =
+    HEAP32[a + ({{{ C_STRUCTS.EmscriptenWebGLContextAttributes.majorVersion }}}>>2)] =
```
Comment (sbc100) [inline src/library_html5_webgl.js:37]: Unrelated whitespace change?

(If this is the result of your editor doing this automatically, I would advise instead binding a key combination to remove trailing whitespace so you are in control over if/when it happens and your PR don't end up containing extra changes like this)

(No need to revert this now.. just a drive by comment)

File: test/other/test_unoptimized_code_size_strict.js.size (hunk 4 above)
Comment (RReverser) [reply test/other/test_unoptimized_code_size_strict.js.size:1]: They are not, as mentioned in description this initially looked like #20530 but actually turned out to be https://github.com/emscripten-core/emscripten/pull/20578 (see that other PR, I'll rebaseline when that fix is in).

File: src/library_html5_webgl.js (hunk 5 above)
Comment (RReverser) [reply src/library_html5_webgl.js:37]: Yeah it is my editor doing it automatically. In most cases it's a useful setting to have by default (so it trims my own spaces w/o me having to remember to do so separately) and it's very rare for projects to have such trailing whitespace, but occasionally it does introduce such extra changes.

Maybe Emscripten could do a sweep do remove all of them at once?

File: test/other/test_unoptimized_code_size_strict.js.size (hunk 4 above)
Comment (sbc100) [reply test/other/test_unoptimized_code_size_strict.js.size:1]: Great

File: src/library_html5_webgl.js (hunk 5 above)
Comment (sbc100) [reply src/library_html5_webgl.js:37]: My advice would still be to not do this automatically... especially if you work on a lot of projects, but I guess you do and your experience has not been bad enough to change that setting :)

I used to have my ediitor setup that way but over time I realized is wanted more control, and I didn't want random extra changes being generated as part of my PRs.. but I guess it depends which projects you work on and how clean they are.

**Comment by sbc100** [comment]
In my PR where I tried to do something like this I ended up doing #19772

```
    if not feature_matrix.caniuse(feature_matrix.Feature.OPTIONAL_CHAINING):
      settings.TRANSPILE_TO_ES5 = True
```

You can test this locally be setting NODE_JS_TEST to v10.19.0.   I have the following in my `.emscripten` so I can switch between different versions of node easily:

```
OLDEST_NODE = '/usr/local/google/home/sbc/bin/node-v10.19.0-linux-x64/bin/node'  
NEW_NODE = '/usr/local/google/home/sbc/bin/node-v19.3.0-linux-x64/bin/node'         
EMSDK_NODE = [os.path.expanduser('~/dev/wasm/emsdk/node/16.20.0_64bit/bin/node')]
                                                                                 
#NODE_JS = NEW_NODE                                                              
#NODE_JS = OLDEST_NODE                                                           
NODE_JS = EMSDK_NODE                                                             
                                                                                 
#NODE_JS_TEST = '/usr/local/google/home/sbc/bin/node-v21.0.0-v8-canary202309143a48826a08-linux-x64/bin/node'
NODE_JS_TEST = EMSDK_NODE                                                        
#NODE_JS_TEST = '/usr/local/google/home/sbc/bin/node-v18.11.0-linux-x64/bin/node'
#NODE_JS_TEST = '/usr/local/google/home/sbc/bin/node-v17.6.0-linux-x64/bin/node' 
#NODE_JS_TEST = '/usr/local/google/home/sbc/bin/node-v15.14.0-linux-x64/bin/node'
#NODE_JS_TEST = '/usr/local/google/home/sbc/bin/node-v16.18.0-linux-x64/bin/node'
#NODE_JS_TEST = [os.path.expanduser('~/bin/node-v18.0.0-v8-canary2022031068abd7e7cd-linux-x64/bin/node')]

JS_ENGINES = [NODE_JS_TEST, V8_ENGINE]
```

Then I just uncomment the NODE_JS_TEST that I want to test with.

File: src/library_html5_webgl.js (hunk 5 above)
Comment (RReverser) [reply src/library_html5_webgl.js:37]: > especially if you work on a lot of projects, but I guess you do and your experience has not been bad enough to change that setting :)

Yeah so far it's been almost always benefitial, with very occasional per-project settings to disable it (like 2 projects in the last couple of years maybe?).

Given that I'm used to working on Rust and JS where automatic formatting is pretty ubiquitous, trimming trailing whitespace has been the least risky thing I had, "auto format on save" can be more problematic 😅 I wish Emscripten enforced autoformatting too though as it takes a lot of unnecessary mental overhead for new contributors by automatically following the existing style of the project. I see it already has configs for both Clang-format and Python, but those are not really enforced.

**Comment by RReverser** [comment]
Oh wow, I never saw that PR, sorry for stepping on your toes with this one. Yeah, using feature matrix for this makes sense.

I could probably also include BCD syntax data in my automatic feature detection PR too.

**Comment by sbc100** [comment]
> Oh wow, I never saw that PR, sorry for stepping on your toes with this one. Yeah, using feature matrix for this makes sense.
> 

No worries!  I half forgot about it myself.. I have so many stale PRs I loose track.


> I could probably also include BCD syntax data in my automatic feature detection PR too.

**Comment by RReverser** [comment]
> > I could probably also include BCD syntax data in my automatic feature detection PR too.

Did you mean to write something after that quote?

**Comment by sbc100** [comment]
> > > I could probably also include BCD syntax data in my automatic feature detection PR too.
> 
> Did you mean to write something after that quote?

Nope

**Comment by RReverser** [comment]
Hm I'm confused but okay 😅

**Comment by RReverser** [comment]
> My comment was inline in the middle of the quote: "No worries! I half forgot about it myself.. I have so many stale PRs I loose track."

Ahhh okay. I thought those were 2 separate quotes, and you replied to the first one, but not to the 2nd one.

**Comment by RReverser** [comment]
FWIW I think TRANSPILE_TO_ES5 strategy generally will need some overhaul... for example, if user sets browser versions high enough it might make sense to transpile only to ES6 so only features like nullish coalescing are converted, but other features can still benefit the end user in terms of code size.

It's a bit change though and I'm not prepared to take it on myself.

**Comment by sbc100** [comment]
> FWIW I think TRANSPILE_TO_ES5 strategy generally will need some overhaul... for example, if user sets browser versions high enough it might make sense to transpile only to ES6 so only post-ES6 features like nullish coalescing are converted, but ES6 features can still benefit the end user in terms of code size. Same for other future versions.
> 
> It's a bit change though and I'm not prepared to take it on myself.

I agree it could be more precise, but I'm also not to worried about these cases... most folks souldn't be transpiling at all, and we tend to favor the default/non-transpile case without focusing too much on the cost (in both build time and code size) of the transpiling case.

**Comment by RReverser** [comment]
Okay, fair enough. Maybe if we ever switch to `swc`, which, like other popular transpilers, uses [browserslist](https://github.com/browserslist/browserslist) to express arbitrary browser conditions like [`"> 0.25%, not dead"`](https://swc.rs/docs/configuration/supported-browsers#targets), we won't need to worry about those MIN_XXX_VERSION as much anyway (only for Wasm features).

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Can this now be re-based and landed?

**Comment by sbc100** [comment]
Looks like a whole bunch of the code size tests are failing too.  Maybe another `emsdk install tot` + clean build + rebase is needed there?

**Comment by sbc100** [comment]
I see a bunch of warnings like this one in the test-node-compat tests:  `emcc: error: enabling transpilation via closure due to browser version settings.  This warning can be suppressed by passing `--closure=1` or `--closure=0` to opt into our explicitly. [-Wtranspile] [-Werror]`.

I would have though that those tests would explicitly disable that warning .. checking..

**Comment by RReverser** [comment]
> Looks like a whole bunch of the code size tests are failing too. Maybe another `emsdk install tot` + clean build + rebase is needed there?

I thought I already did that during rebase - I certainly reran all code size tests - but maybe missed one of the cleanup steps. I'll go through that again once the actual issues are solved.

**Comment by RReverser** [comment]
> I would have though that those tests would explicitly disable that warning .. checking..

Ah yeah, that's why they're failing. Disabling the warning makes sense I think.

**Comment by sbc100** [comment]
My guess is that we somehow need to inject `-Wno-transpile` when running tests against the old version of node in test-node-compat.

Perhaps the test runner could do this automatically?   Or you could do some thing really gross like `EMCC_CFLAGS=-Wno-transpile` in the circleci config.

**Comment by RReverser** [comment]
Ah I think I found it - it comes from `building.get_emcc_node_flags(shared.check_node_version())`?

If so, disabling this warning in `test/common.py` in the same place where it's used seems like the best place.

**Review by sbc100 — APPROVED** [review]
Great!

File: src/library_wget.js [hunk 6]
```diff
@@ -232,9 +232,7 @@ var LibraryWget = {
   emscripten_async_wget2_abort__proxy: 'sync',
   emscripten_async_wget2_abort: (handle) => {
     var http = wget.wgetRequests[handle];
-    if (http) {
-      http.abort();
-    }
+    http?.abort();
```
Comment (sbc100) [inline src/library_wget.js:235]: Maybe combine these into a single line and remove the curlies?

Comment (RReverser) [reply src/library_wget.js:235]: It's an easy change but I'd prefer to leave unrelated cleanups separate.

This is already a pretty large PR and as mentioned above I already had some mistakes slipped in it before so I'd rather 1) have someone do another pass to check if conditions look good and 2) not extend it with more cleanups.

**Comment by RReverser** [comment]
Hm now just `test_node_unhandled_rejection` is failing, with empty output. Not sure why.

**Comment by sbc100** [comment]
> Hm now just `test_node_unhandled_rejection` is failing on Node compat test, with empty output. Not sure why.

That test is supposed to fail (i.e. node should return non-zero).  Can your repro locally?

**Comment by sbc100** [comment]
> FWIW the only relevant change that I can think of is that that test will now run with transpilation. I see it uses async-await, which could behave differently when transpiled by Closure I guess, especially in terms of edge cases like reporting rejections.

Can you repro the issue locally if you run add `-sLEGACY_VM_SUPPORT` to the test?

**Comment by RReverser** [comment]
> Can you repro the issue locally if you run add `-sLEGACY_VM_SUPPORT` to the test?

Yup. Here's the generated code after transpilation: https://gist.github.com/RReverser/61be1e139a7b672dbb2edfcbbfcb1335

It's interesting that there is some Closure variable `FORCE_POLYFILL_PROMISE_WHEN_NO_UNHANDLED_REJECTION` in there. I guess it's relevant.

**Comment by RReverser** [comment]
Ah no that was a misdirect.

What actually seems to happen is that in

```js
    async function foo() {
      var a = missing;
    }
```

part of the test Closure Compiler outright ignores / removes this part. Idk if that's because `missing` is not declared or because `a` is not used, but either way the generated async function doesn't do anything and immediately successfully returns: https://gist.github.com/RReverser/61be1e139a7b672dbb2edfcbbfcb1335#file-output-js-L773-L777

```js
function foo() {
  return $jscomp.asyncExecutePromiseGeneratorProgram(function(a) {
    a.jumpToEnd();
  });
}
```

This looks like a Closure bug but I don't want to dig deeper into it and instead will replace contents with explicit `throw`

```js
    async function foo() {
      throw new Error("failed");
    }
```

so that it can do its original purpose of testing for unhandled rejections.

**Review by sbc100 — APPROVED** [review]

**Comment by RReverser** [comment]
So close... but `other.test_closure_type_annotations` looks like a real failure. Comparing output JS saved via `--save-dir`, it does look like a lot of code that was previously DCEd is now retained.

I'm not sure what's causing it yet.

**Comment by RReverser** [comment]
Doesn't solve the issue, but this construction in the test looks odd btw: https://github.com/emscripten-core/emscripten/blob/5137bfc0470d72d95869c50b25eb87311735471c/test/test_other.py#L9274-L9275

`assertFalse` takes an expression as its first argument, but this is passing a message there instead. `if expr` on the message still works, but maybe a better way to write it would be

```py
self.assertFalse('my ' + attribute in code, 'Attribute `%s` could not be DCEd' % attribute)
```

**Comment by sbc100** [comment]
> Doesn't solve the issue, but this construction in the test looks odd btw:
> 
> https://github.com/emscripten-core/emscripten/blob/5137bfc0470d72d95869c50b25eb87311735471c/test/test_other.py#L9274-L9275
> 
> `assertFalse` takes an expression as its first argument, but this is passing a message there instead. `if expr` on the message still works, but maybe a better way to write it would be
> 
> ```python
> self.assertFalse('my ' + attribute in code, 'Attribute `%s` could not be DCEd' % attribute)
> ```

Probably historical reasons.

And using f-strings would be even better.

**Comment by RReverser** [comment]
> I'm not sure what's causing it yet. That test shouldn't go via transpiling, so that's not the cause, but the only other change is usage of `?.` operators... maybe Closure can't optimise them yet?

Testing a bit further and comparing diffs with `--closure-args=--debug` added to the test, this seems like the problem. Some slight change in semantics is causing Closure to keep all this code.

That's pretty unfortunate and I don't see where to go from here, other than maybe reporting issue to them.

**Comment by sbc100** [comment]
I don't quite understand the reason that `other.test_closure_type_annotations` would start failing?  Is it due to the change made to the JS library input code?

File: tools/link.py [hunk 7]
```diff
@@ -1127,18 +1127,22 @@ def phase_linker_setup(options, state, newargs):
...
-                                 settings.MIN_FIREFOX_VERSION < 44 or
-                                 settings.MIN_CHROME_VERSION < 49 or
-                                 settings.MIN_SAFARI_VERSION < 110000 or
+    #                             Max Version: EDGE:85 FF:79 CHROME:85 SAFARI:14 NODE:16
+    # TODO: replace this with feature matrix in the future.
+    settings.TRANSPILE_TO_ES5 = (settings.MIN_EDGE_VERSION < 85 or
+                                 settings.MIN_FIREFOX_VERSION < 79 or
+                                 settings.MIN_CHROME_VERSION < 85 or
+                                 settings.MIN_SAFARI_VERSION < 140000 or
+                                 settings.MIN_NODE_VERSION < 160000 or
```
Comment (sbc100) [inline tools/link.py:1145]: Maybe you can land this change on its own / separately?

**Comment by RReverser** [comment]
> Is it due to the change made to the JS library input code?

Yeah looks like at least some of the changes cause Closure to retain variables that would be previously DCEd. Looking at the diff, those are huge chunks of SDL, WebGL and HTML5 API bindings that were previously removed. But there's nothing special about changes to those libraries in particular...

**Comment by RReverser** [comment]
Another offtopic, but you landed that PR with automatic CI retries on failure, right? Do you know why those flaky tests are still failing then?

**Comment by sbc100** [comment]
> Another offtopic, but you landed that PR with automatic CI retries on failure, right? Do you know why those flaky tests are still failing then?

There are two mechanisms for re-trying... tests marked a `@flakey` explictly are re-tried if they fail.  We have very few such tests and non of the browser tests are yet marked in this way.

However because browser tests are just generally more flakey we have a second mechanism that retries all browser tests then they fail. 

However, I believe the most of the failures we are seeing in the browser test suite today involve the browser crashing or otherwise becoming unresponsive.. which then results in all subsequest browser tests failing with "[no http server activity]".   I don't believe any amount of re-trying will help in this case.  I think we need to somehow restart the browser when this happens, but we don't currently have a mechanism for this.  CC @kripken

**Comment by RReverser** [comment]
Huh, that's... not at all the kind of changes I expected to be problematic.

```diff
diff --git a/src/library_browser.js b/src/library_browser.js
index 67accc101..6e5904b1a 100644
--- a/src/library_browser.js
+++ b/src/library_browser.js
@@ -736,7 +736,7 @@ var LibraryBrowser = {
     {{{ runtimeKeepalivePush() }}}
 
     var _suffix = UTF8ToString(suffix);
-    Browser.asyncPrepareDataCounter ||= 0;
+    if (!Browser.asyncPrepareDataCounter) Browser.asyncPrepareDataCounter = 0;
     var name = 'prepare_data_' + (Browser.asyncPrepareDataCounter++) + '.' + _suffix;
     var cname = stringToNewUTF8(name);
     FS.createPreloadedFile(
diff --git a/src/library_webgl.js b/src/library_webgl.js
index b4cde3c73..6c26b45fc 100644
--- a/src/library_webgl.js
+++ b/src/library_webgl.js
@@ -231,7 +231,9 @@ for (/**@suppress{duplicate}*/var i = 0; i < {{{ GL_POOL_TEMP_BUFFERS_SIZE }}};
     // cleared the stored error by a call to glGetError().
     recordError: function recordError(errorCode) {
 #if GL_TRACK_ERRORS
-      GL.lastError ||= errorCode;
+      if (!GL.lastError) {
+        GL.lastError = errorCode;
+      }
 #endif
     },
     // Get a new ID for a texture/buffer/etc., while keeping the table dense and
```

Reverting just those 2 changes is enough to make test pass. This feels almost random :/

**Comment by sbc100** [comment]
> Huh, that's... not at all the kind of changes I expected to be problematic.
> 
> ```diff
> diff --git a/src/library_browser.js b/src/library_browser.js
> index 67accc101..6e5904b1a 100644
> --- a/src/library_browser.js
> +++ b/src/library_browser.js
> @@ -736,7 +736,7 @@ var LibraryBrowser = {
>      {{{ runtimeKeepalivePush() }}}
>  
>      var _suffix = UTF8ToString(suffix);
> -    Browser.asyncPrepareDataCounter ||= 0;
> +    if (!Browser.asyncPrepareDataCounter) Browser.asyncPrepareDataCounter = 0;
>      var name = 'prepare_data_' + (Browser.asyncPrepareDataCounter++) + '.' + _suffix;
>      var cname = stringToNewUTF8(name);
>      FS.createPreloadedFile(
> diff --git a/src/library_webgl.js b/src/library_webgl.js
> index b4cde3c73..6c26b45fc 100644
> --- a/src/library_webgl.js
> +++ b/src/library_webgl.js
> @@ -231,7 +231,9 @@ for (/**@suppress{duplicate}*/var i = 0; i < {{{ GL_POOL_TEMP_BUFFERS_SIZE }}};
>      // cleared the stored error by a call to glGetError().
>      recordError: function recordError(errorCode) {
>  #if GL_TRACK_ERRORS
> -      GL.lastError ||= errorCode;
> +      if (!GL.lastError) {
> +        GL.lastError = errorCode;
> +      }
>  #endif
>      },
>      // Get a new ID for a texture/buffer/etc., while keeping the table dense and
> ```
> 
> Reverting just those 2 changes is enough to make test pass. This feels almost random :/

I little worrying.. but this change lgtm still

**Comment by RReverser** [comment]
> but this change lgtm still

Hm are you sure, despite the bloat it potentially adds?

**Comment by RReverser** [comment]
Should I remove the test then or?

**Comment by RReverser** [comment]
> to DCE other unrelated stuff

Somewhat related stuff (e.g. that `GL.lastError` change seems to keep parts of WebGL that the tests checks again), but yeah still WTF is the right reaction.

> I think you should just revert those parts of the change.

FWIW even without them manual verification shows that the output JS is still much larger than before this PR. It's just that it passes the specific checks in the test.

**Comment by RReverser** [comment]
Trying with Closure, it seems it treats those operators somewhat differently when using globals.

Standalone sample:

```js
globalThis.foo = () => {
  if (!GL.lastError) {
    GL.lastError = errorCode;
  }
};

globalThis.bar = () => {
  GL.lastError ||= errorCode;
}
```

Pretty-printed output after advanced opts:

```js
globalThis.h = function() {
  GL.lastError || (GL.lastError = errorCode);
};
globalThis.g = function() {
  var a;
  (a = GL).lastError || (a.lastError = errorCode);
};
```

I think Closure is "worried" that the globals might happen to be getters on the global object, in which case they might have side effects, so they need to be stored in a separate variable first, which breaks some optimisations.

If this assumption is correct, it would explain why the issue occurs only when accessing properties on our own `Browser` and `GL` variables, which become globals in the default Emscripten's output.

I wonder if there's some way to annotate those variables as safe to access for Closure.

**Comment by sbc100** [comment]
> Trying with Closure, it seems it treats those operators somewhat differently when using globals.
> 
> Standalone sample:
> 
> ```js
> globalThis.foo = () => {
>   if (!GL.lastError) {
>     GL.lastError = errorCode;
>   }
> };
> 
> globalThis.bar = () => {
>   GL.lastError ||= errorCode;
> }
> ```
> 
> Pretty-printed output after advanced opts:
> 
> ```js
> globalThis.h = function() {
>   GL.lastError || (GL.lastError = errorCode);
> };
> globalThis.g = function() {
>   var a;
>   (a = GL).lastError || (a.lastError = errorCode);
> };
> ```
> 
> I think Closure is "worried" that the globals might happen to be getters on the global object, in which case they might have side effects, so they need to be stored in a separate variable first to ensure they're accessed only once when using `||=`, and that, in turn, breaks some optimisations.
> 
> If this assumption is correct, it would explain why the issue occurs only when accessing properties on our own `Browser` and `GL` variables, which become globals in the default Emscripten's output.
> 
> I wonder if there's some way to annotate those variables as safe to access for Closure.

This makes me wonder if we need more code size tests to catch this kind of thing.. perhaps one that does `-sINCLUDE_FULL_LIBRARY`?

How about landing the just the functional part of this change first without that actual mass usage of the new operators?  (Or perhaps just one usage).   Then the large scale change can follow once we figure out the closure stuff?

**Comment by RReverser** [comment]
> perhaps one that does `-sINCLUDE_FULL_LIBRARY`?

That's what the failing test does at the moment. Although maybe a size test for `INCLUDE_FULL_LIBRARY` would also be good.

> How about landing the just the functional part of this change first without that actual mass usage of the new operators?

Main functional part has already landed in https://github.com/emscripten-core/emscripten/pull/20549, the only remaining functional bit is transpilation versions but it's pretty tiny.

I first want to see if it's possible to either annotate those few vars somehow or just revert those files in particular for now.

I tried adding `$Browser__docs: '/** @const */',` but that reveals a bunch of other issues with function signature mismatch and whatnot inside `library_browser.js`, so looking for a less impactful solution for now.

**Comment by RReverser** [comment]
Ok I applied just the simple fix to those methods. Idk what to do about Closure issue in general, but since you're okay with it and the increased size of INCLUDE_FULL_LIBRARY is not covered by tests, I suppose it should be fine.
